testing/integration/datastreams_test.go

```diff
@@ -0,0 +1,400 @@
+	if err != nil {
+		t.Fatalf("could not create new fixture: %s", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// 1. Create a policy in Fleet with monitoring enabled.
+	// To ensure there are no conflicts with previous test runs against
+	// the same ESS stack, we add the current time at the end of the policy
+	// name. This policy does not contain any integration.
+	createPolicyReq := kibana.AgentPolicy{
+		Name:        t.Name() + "--" + time.Now().Format(time.RFC3339Nano),
+		Namespace:   info.Namespace,
+		Description: "Test policy for " + t.Name(),
+		MonitoringEnabled: []kibana.MonitoringEnabledOption{
+			kibana.MonitoringEnabledLogs,
+			kibana.MonitoringEnabledMetrics,
+		},
+		IsProtected: false,
+	}
+	installOpts := atesting.InstallOpts{
+		NonInteractive: true,
+		Force:          true,
+	}
+
+	// 2. Install the Elastic-Agent with the policy that
+	// was just created.
+	policy, err := tools.InstallAgentWithPolicy(ctx,
+		t,
+		installOpts,
+		agentFixture,
+		info.KibanaClient,
+		createPolicyReq)
+	if err != nil {
+		t.Fatalf("could not install Elastic-AGent with Policy: %s", err)
+	}
+
+	// 3. Prepare a request to add an integration to the policy
+	tmpl, err := template.New(t.Name() + "custom-log-policy").Parse(policyJSON)
+	if err != nil {
+		t.Fatalf("cannot parse template: %s", err)
+	}
+
+	// The time here ensures there are no conflicts with the integration name
+	// in Fleet.
+	agentPolicyBuffer := bytes.Buffer{}
+	err = tmpl.Execute(&agentPolicyBuffer, plolicyVars{
+		Name:        "Log-Input-" + t.Name() + "-" + time.Now().Format(time.RFC3339),
+		PolicyID:    policy.ID,
+		LogFilePath: logFilePath,
+		Namespace:   dsNamespace,
+		Dataset:     dsDataset,
+	})
+	if err != nil {
+		t.Fatalf("could not render template: %s", err)
+	}
+
+	// 4. Call Kibana to create the policy.
+	// Docs: https://www.elastic.co/guide/en/fleet/current/fleet-api-docs.html#create-integration-policy-api
+	resp, err := info.KibanaClient.Connection.Send(
+		http.MethodPost,
+		"/api/fleet/package_policies",
+		nil,
+		nil,
+		&agentPolicyBuffer)
+	if err != nil {
+		t.Fatalf("could not execute request to Kibana/Fleet: %s", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		// On error dump the whole request response so we can easily spot
+		// what went wrong.
+		t.Errorf("received a non 200-OK when adding package to policy. "+
+			"Status code: %d", resp.StatusCode)
+		respDump, err := httputil.DumpResponse(resp, true)
+		if err != nil {
+			t.Fatalf("could not dump error response from Kibana: %s", err)
+		}
+		t.Log("Kibana error response")
+		t.Log(string(respDump))
+		t.FailNow()
+	}
+
+	require.Eventually(
+		t,
+		ensureDocumentsInES(t, context.TODO(), info.ESClient, dsType, dsDataset, dsNamespace, int(numEvents)),
+		120*time.Second,
+		time.Second,
+		"could not get all expected documents form ES")
+}
+
+func TestFlattenedDatastreamStandalone(t *testing.T) {
```

Comment:
1. Do we need an entire integration test to test this? Why do we need an Elastic stack? Could this be tested with the fake component in a unit test?
2. Why do we need to test this for both fleet and standalone? What actually differs between these cases that we need to cover them separately? Does how the policy arrives at the agent change how it gets parsed?

Reply:
> Why do we need to test this for both fleet and standalone? What actually differs between these cases that we need to cover them separately?

I'm being extra cautious here. When I first implemented it in Beats, all my test cases were passing, but when running Elastic-Agent managed by Fleet, some fields were set slightly different that my handcrafted standalone policy, which broke all other cases.

Learning from that mistake I thought it would be good to have:
1. A test case managed by Fleet because the change affects how the policy is parsed/interpreted
2. Something that ensures we're not breaking the normal behaviour the Elastic-Agent is run managed by Fleet and a simple integration is added/used.

> Does how the policy arrives at the agent change how it gets parsed?

I'm not quite sure I fully understood this question. Depending on how the policy is sent to Elastic-Agent it can arrive in the `DataStream` field from the input/streams or in the `Source` field as YAML that needs to be decoded/parsed. The test cases should cover a mix of both and the "default behaviour" of just installing Elastic-Agent and adding integrations without changing any of the defaults.

> Do we need an entire integration test to test this? Why do we need an Elastic stack? Could this be tested with the fake component in a unit test?

I believe so, I wanted to test it end-to-end, so ensuring that the Elastic-Agent can properly interpret the `data_stream` configuration no matter how it's sent but also that Beats will work as expected when receiving the information from Elastic-Agent. That is also why I check for the documents on ES with the expected index name rather than asserting it any other way.

Probably the Fleet managed test could be renamed to reflect something like "simple happy path always works".

Reply:
We are learning that maintaining integration tests is very expensive, so I only want us to write and maintain integration tests for features that can only be tested this way. I don't want us to have integration tests because it was possible, or "just in case" :) Our test suite will take hours and be difficult to maintain if we do this everywhere.

I would not expect the agent policy to be parsed differentlu when sent by Fleet. The part of the agent that converts the policy to components and parses it is centralized into the coorindator, not the configuration providers. See https://github.com/elastic/elastic-agent/blob/main/docs/architecture.md#coordinator

What I would expect is that the Fleet will perhaps use different variants of the syntax than you may have expected when writing standalone configurations by hand.

I could be convinced we need a Fleet integration test for this functionality, but I'm not convinced that we need an entirely new test for it. Could we just extend the monitoring and log test in https://github.com/elastic/elastic-agent/blob/main/testing/integration/monitoring_logs_test.go to also install the custom logs integration? That test is already exactly the same as the one you added minus that one integration. This would avoid the cost of creating an maintaining an entirely new test for a few small additions to an existing tests. Integration tests are not unit tests, they don't need to test only one piece of functionality.

I'm not convinced we need a standalone test case since there should not be a difference between Fleet and standalone agents in this case. If we need to cover all possible permutations of the YAML syntax we should be doing that in unit tests.

Reply:
I see your point and agree in not creating integrations test "just in case". In this specific case both, the Feet managed and standalone, tests also test the most basic use case for Elastic-Agent and Beats: log ingestion. I believe it is valuable to have at least one test like this for each case. They might be more like a smoke test but I still believe they're valuable.

I agree with not creating an entire new test for it. The way I coded both tests they're very isolated in the sense that all documents ingested go to indexes different than the default ones. I can refactor them to use the same stack deployment as `TestMonitoringLogsShipped`.

Reply:
>  I can refactor them to use the same stack deployment as TestMonitoringLogsShipped.

👍 

I still don't think we need both a standalone and a fleet managed test for this, only one of them is enough. There is no functional difference besides how the policy is delivered to the agent. If you were to copy the agent policy generated by Fleet into a standalone agent the behaviour would be exactly the same.

Reply:
> If you were to copy the agent policy generated by Fleet into a standalone agent the behaviour would be exactly the same.

While that is true, the opposite is not. At least regarding the characters allowed for the `data_stream.*` fields. Fleet server/Kibina is more picky about it than a standalone Elastic-Agent.

If you really want I can remove the standalone test. I tried to re-use the same stack+VM, however the Elastic-Agent wasn't being properly uninstalled and I didn't want to spend time debugging it. I'd be happy to go back to it and make sure the same stack+VM can run all tests.

From my machine, with an existing stack and VM, the standalone Go test takes about 20s and the whole execution of `mage integration:single <test name>` takes about 50s.
```
>>> (linux-amd64-ubuntu-2204) Test output (sudo) (stdout): --- PASS: TestFlattenedDatastreamStandalone (19.61s)
>>> (linux-amd64-ubuntu-2204) Test output (sudo) (stdout): PASS
>>> (linux-amd64-ubuntu-2204) Test output (sudo) (stdout): ok   github.com/elastic/elastic-agent/testing/integration    19.624s
>>> (linux-amd64-ubuntu-2204) Test output (sudo) (stdout): DONE 1 tests in 24.121s
>>> (linux-amd64-ubuntu-2204) Test output (sudo) (stdout): >> go test: remote-linux-amd64-ubuntu-2204-sudo.integration Test Passed
>>> Testing completed (1 successful)
>>> Console output written here: build/TEST-go-integration.out
>>> Console JSON output written here: build/TEST-go-integration.out.json
>>> JUnit XML written here: build/TEST-go-integration.xml
>>> Diagnostic output (if present) here: build/diagnostics
AGENT_KEEP_INSTALLED=false TEST_PLATFORMS=linux/amd64 mage -v    3.58s user 1.51s system 10% cpu 48.286 total
```

Reply:
> Fleet server/Kibina is more picky about it than a standalone Elastic-Agent.

It's not clear to me what this means. Do you have an example?

If Fleet and Standalone agents have different views on what syntax is valid that is a bug. The two should be aligned. The allowed syntax must be the same. The actual syntax variant used may differ, but any form should be valid in either place.

Reply:
What error do you get from Fleet when you include the `-`?

It should be valid for an index name to contain a `-` going off of https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html#indices-create-api-path-params.

This smells like Fleet is unnecessarily strict, that doesn't justify an entirely separate agent integration test to me, but we'll confirm first.

Reply:
If I make a request like this:
```
POST kbn:/api/fleet/package_policies
{
  "policy_id": "cfbf25e0-68ee-11ee-8f57-2948faf178f3",
  "package": {
    "name": "log",
    "version": "2.3.0"
  },
  "name": "foo-logs",
  "namespace": "my-namespace",
  "inputs": {
    "logs-logfile": {
      "enabled": true,
      "streams": {
        "log.logs": {
          "enabled": true,
          "vars": {
            "paths": [
              "/tmp/foo.ndjson"
            ],
            "data_stream.dataset": "my-dataset"
          }
        }
      }
    }
  }
}
```

Kibana/Fleet returns:
```
{
  "statusCode": 500,
  "error": "Internal Server Error",
  "message": "[namespace]: Namespace contains invalid characters"
}
```

If I try to put it in the input or stream level I get a Bad Request:
```
POST kbn:/api/fleet/package_policies
{
  "policy_id": "cfbf25e0-68ee-11ee-8f57-2948faf178f3",
  "package": {
    "name": "log",
    "version": "2.3.0"
  },
  "name": "foo-logs",
  "inputs": {
    "logs-logfile": {
      "enabled": true,
      "data_stream": {
        "namespace": "my-namespace"
      },
      "streams": {
        "log.logs": {
          "enabled": true,
          "vars": {
            "paths": [
              "/tmp/foo.ndjson"
            ],
            "data_stream.dataset": "my-dataset"
          }
        }
      }
    }
  }
}
```
Response
```
{
  "statusCode": 400,
  "error": "Bad Request",
  "message": """[request body]: types that failed validation:
- [request body.0.inputs]: expected value of type [array] but got [Object]
- [request body.1.inputs.logs-logfile.data_stream]: definition for this key is missing"""
}
```

Reply:
If you really don't want both tests, I can remove one of them. Do you have any preferences?

I tried to have both, the standalone and Fleet managed using the same Stack/VM, but the Elastic-Agent managed by fleet is not being correctly stopped/uninstalled despite no error from the framework, so the standalone agent won't start :/

Reply:
Found the reason Fleet is not allowing the dash, it is "forbidden" in the data stream naming scheme: https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme

> In all three parts, the “-” character is not allowed. This means all data streams are named in the following way:

Also enforcing this in standalone agents would be ideal, but that is a breaking change for anyone that has done this already. I would just note this as a comment.

> If you really don't want both tests, I can remove one of them. Do you have any preferences?

The Fleet test is the one most likely to catch actual problems with the index names, as we have discovered so let's keep that one.